### ISSUE TITLE
False positive on `UnusedPrivateMember` for actual/expected inline method parameters.

### ISSUE THREAD

**Issue by Whathecode** [detekt/detekt#3415]
## Expected Behavior
Expected inline methods in a Kotlin multiplatform project should not give `UnusedPrivateMember` warning for parameters.

## Observed Behavior/Steps to Reproduce
```
internal expect object Reflection
{
    internal inline fun <reified T> extendsType( klass: KClass<*> ): Boolean
}
```

This gives a `UnusedPrivateMember` warning for the `klass` parameter.

## Your Environment

* Version of detekt used: 1.16.0-RC1
* Version of Gradle used (if applicable): 6.6
* Operating System and version: Windows 10
* Link to your project (if it's a public repository): https://github.com/cph-cachet/carp.core-kotlin/blob/develop/carp.common/src/commonMain/kotlin/dk/cachet/carp/common/reflect/ReflectionAvailable.kt#L32

**Comment by cortinico** [comment]
Thanks for the report. The `expect object` scenario is indeed not covered by this rule. Fix is in #3417

**Comment by Whathecode** [comment]
@schalkms I just updated to 1.16.0-RC2 This is now fixed for the `expect` definition, but still gives the same warning on my actual implementation.

```
internal actual object Reflection
{
    internal actual inline fun <reified T> extendsType( klass: KClass<*> ): Boolean =
        throw UnsupportedOperationException()
}
```

**Comment by cortinico** [comment]
While working on this I noticed that we have a misalignment between `expect` and `actual` and it seems to be intentional: https://github.com/detekt/detekt/pull/2643#discussion_r416774303

@schalkms is that still relevant?

**Comment by schalkms** [comment]
I don't think it's intentional. The fix incorporated for this issue should include the check for the `actual` definition. There seems to be a misalignment between the `actual` and `expect` in the rule implementation.

https://github.com/detekt/detekt/blob/c5f65a9c8ed029f0382674371c070ec0f3071b28/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt#L204

**Comment by cortinico** [comment]
> There seems to be a misalignment between the `actual` and `expect` in the rule implementation.

Yup that's exactly why I asked. In your comment here: https://github.com/detekt/detekt/pull/2643#discussion_r416774303 you mentioned:

> Actual classes are an implementation concern. Thus, unused parameters should be reported.

I'd like to understand more about that. If we don't report `UnusedPrivateMember` in the `expect` declaration but we do so for the `actual`, how are users supposed to fix it?

**Comment by schalkms** [comment]
@cortinico thanks for asking the right questions! 🙂
I looked at my notes. Now I can remember it. `Actual` is an implementation detail, meaning such a function has a body in contrast to an `expect` function. Hence, the unused parameter should be reported by this rule for an `actual` function.

> With this mechanism, a common source set defines an expected declaration, and platform source sets must provide the actual declaration that corresponds to the expected declaration. 

https://kotlinlang.org/docs/multiplatform.html

**Comment by cortinico** [comment]
> Hence, the unused parameter should be reported by this rule for an `actual`, but not for an `expect` function.

Agree. Unless we have a strong opinion here or a valid use case for changing it, I'm leaning towards re-closing this issue. 

> but still gives the same warning on my actual implementation.

@Whathecode does this clarifies your question?

**Comment by Whathecode** [comment]
> @Whathecode does this clarifies your question?

Well, it wasn't really a question. More of a continued bug report. :)

> Hence, the unused parameter should be reported by this rule for an actual, but not for an expect function.

The problem is if you rename the, correctly reported as 'unused', parameter to `_`, you get a complaint that the actual implementation does not match the expected implementation. Thus, this should _also_ be ignored for actual implementations. 

But, ... it is possible I overlooked something. Only looked at this shortly and just slammed the `@Suppress` back on there.

**Comment by cortinico** [comment]
> Thus, this should also be ignored for actual implementations.

Can we bring someone with more KMM expertise to drop a suggestion here?

**Comment by sczerwinski** [comment]
@cortinico,

As someone with some experience in Kotlin Multiplatform projects, I would say:
- if a rule is ignored for `abstract` members, it should be ignored for `expect` members,
- if a rule is ignored for **overridden** members, it should be ignored for `actual` members.

Basically, `expect` class/member is a placeholder defined in common sources, which should be implemented as a separate `actual` class/member for every defined platform.

Just like with abstract and overridden member functions, a parameter can be used in one `actual` implementation, but not the other.

See also:
- #3388
- #3390
